File: 2020/day04/day04.jl
mutable struct Passport
    byr::String
    iyr::String
    eyr::String
    hgt::String
    hcl::String
    ecl::String
    pid::String
    cid::String
    Passport() = new("","","","","","","","")
    Passport(byr,iyr,eyr,hgt,hcl, ecl, pid, cid) =
        new(byr,iyr,eyr,hgt,hcl, ecl, pid, cid)
end

function is_valid(pass::Passport)::Bool
    for i in 1:(length(fieldnames(Passport))-1)
        if isempty(getfield(pass, i))
            return false
        end
    end
    return true
end

function is_valid2(pass::Passport)::Bool
    if !is_valid(pass)
        return false
    end

    # check birth, issue and expiration year
    if !(1920 <= parse(Int64, pass.byr) <= 2002) ||
        !(2010 <= parse(Int64, pass.iyr) <= 2020) ||
        !(2020 <= parse(Int64, pass.eyr) <= 2030)
        return false
    end

    # check height
    units = pass.hgt[end-1:end]
    if units == "cm"
        if !(150 <= parse(Int64, pass.hgt[1:end-2]) <= 193)
           return false
        end
    elseif units == "in"
        if !(59 <= parse(Int64, pass.hgt[1:end-2]) <= 76)
            return false
        end
    else
        return false
    end

    # check hair color
    if pass.hcl[1] != '#' || length(pass.hcl) != 7
        return false
    end

    for h in pass.hcl[2:end]
        if !('0' <= h <= '9') && !('a' <= h <= 'f')
            return false
        end
    end

    # check eye color
    avail_color = ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]
    if !(pass.ecl in avail_color)
        return false
    end

    # check passport ID
    if length(pass.pid) != 9 || !any(isdigit, pass.pid)
       return false
    end

    return true
end

function main(filename::String)
    lines = ""
    for l in eachline(filename, keep=true)
        lines *= l
    end

    lines = replace(replace(replace(lines,"\n\n"=>"+"), "\n"=>" "), "+"=>'\n')

    lines = split(lines, '\n')

    passports = Vector{Passport}(undef, length(lines))
    for (i, line) in enumerate(lines)
        part_line = split(line, ' ')

        temp_pass = Passport()
        for part in part_line
            if findfirst("byr", part) !== nothing
                temp_pass.byr = String(part[5:end])
            elseif findfirst("iyr", part) !== nothing
                temp_pass.iyr = String(part[5:end])
            elseif findfirst("eyr", part) !== nothing
                temp_pass.eyr = String(part[5:end])
            elseif findfirst("hgt", part) !== nothing
                temp_pass.hgt = String(part[5:end])
            elseif findfirst("hcl", part) !== nothing
                temp_pass.hcl = String(part[5:end])
            elseif findfirst("ecl", part) !== nothing
                temp_pass.ecl = String(part[5:end])
            elseif findfirst("pid", part) !== nothing
                temp_pass.pid = String(part[5:end])
            elseif findfirst("cid", part) !== nothing
                temp_pass.cid = String(part[5:end])
            end
        end

        passports[i] = temp_pass
    end


    println("Part1: ", length(filter(is_valid, passports)))
    println("Part2: ", length(filter(is_valid2, passports)))

end

File: 2020/day04/day04.cpp
#include "../../utils.hpp"
#include <sstream>

using namespace std;
using namespace utils;

struct Passport {
  string_view byr, iyr, eyr;
  string_view hgt;
  string_view hcl, ecl;
  string_view pid, cid;

  bool is_present() const {
    return !byr.empty() && !iyr.empty() && !eyr.empty() && !hgt.empty() &&
           !hcl.empty() && !ecl.empty() && !pid.empty();
  }

  bool is_valid() const {
    if (!is_present()) {
      return false;
    }

    if (!inRange(byr, 1920, 2002) || !inRange(iyr, 2010, 2020) ||
        !inRange(eyr, 2020, 2030)) {
      return false;
    }

    bool checkHeight = false;
    if (hgt.find("cm") != -1) {
      checkHeight = inRange(hgt.substr(0, hgt.size() - 2), 150, 193);
    }

    if (hgt.find("in") != -1) {
      checkHeight = inRange(hgt.substr(0, hgt.size() - 2), 59, 76);
    }

    if (!checkHeight) {
      return false;
    }

    if (hcl[0] != '#' || hcl.size() != 7 ||
        hcl.substr(1).find_first_not_of("0123456789abcdef") != -1) {
      return false;
    }

    if (!match(ecl, "amb", "blu", "brn", "gry", "grn", "hzl", "oth")) {
      return false;
    }

    if (pid.size() != 9 || !try_int(pid).has_value) {
      return false;
    }

    return true;
  }
};

int main(int argc, char *argv[]) {

  if (argc == 1) {
    cerr << "No file given!\n";
    exit(1);
  }

  auto file = read_file_as_string_view(argv[1]);
  auto splitter = [](const auto &x) { return split_by(x, '\n'); };
  auto lines = map(splitter, split_by(file, "\n\n"));

  vector<Passport> passports;
  passports.reserve(lines.size());

  for (auto &l : lines) {
    Passport pass = {};
    auto line = join_string_view(l);
    auto parts = split_by(line, ' ');

    for (auto &p : parts) {
      if (p.find("byr:") == 0) {
        pass.byr = p.substr(4);
      } else if (p.find("iyr:") == 0) {
        pass.iyr = p.substr(4);
      } else if (p.find("eyr:") == 0) {
        pass.eyr = p.substr(4);
      } else if (p.find("hgt:") == 0) {
        pass.hgt = p.substr(4);
      } else if (p.find("hcl:") == 0) {
        pass.hcl = p.substr(4);
      } else if (p.find("ecl:") == 0) {
        pass.ecl = p.substr(4);
      } else if (p.find("pid:") == 0) {
        pass.pid = p.substr(4);
      } else if (p.find("cid:") == 0) {
        pass.cid = p.substr(4);
      }
    }
    passports.push_back(pass);
  }

  auto part1 =
      filter([](auto &p) -> bool { return p.is_present(); }, passports).size();

  auto part2 =
      filter([](auto &p) -> bool { return p.is_valid(); }, passports).size();

  cout << "Part1: " << part1 << '\n';
  cout << "Part2: " << part2 << '\n';

  return 0;
}
